### PR TITLE
Implement Parentheses Generator using Backtracking Strategy

### PR DESCRIPTION
**Pull request by TruongNhanNguyen**
## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I ran bellow commands using the latest version of **rust nightly**.
- [x] I ran `cargo clippy --all -- -D warnings` just before my last commit and fixed any issue that was found.
- [x] I ran `cargo fmt` just before my last commit.
- [x] I ran `cargo test` just before my last commit and all tests passed.
- [x] I added my algorithm to the corresponding `mod.rs` file within its own folder, and in any parent folder(s).
- [x] I added my algorithm to `DIRECTORY.md` with the correct link.
- [x] I checked `COUNTRIBUTING.md` and my code follows its guidelines.

### CODE REVIEW

File: src/backtracking/parentheses_generator.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+/// * `result` - A mutable reference to the vector storing the generated combinations.
+fn generate(current: &str, open: u32, close: u32, n: u32, result: &mut Vec<String>) {
+    if current.len() == (n * 2) as usize {
+        result.push(current.to_string());
+        return;
+    }
+
+    if open < n {
+        let mut new_str = current.to_string();
+        new_str.push('(');
```
Comment (vil02) [inline src/backtracking/parentheses_generator.rs:36]:
```suggestion
        let new_str = current.to_string() + "(";
        new_str.push('(');
```

File: src/backtracking/parentheses_generator.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        generate(&new_str, open + 1, close, n, result);
+    }
+
+    if close < open {
+        let mut new_str = current.to_string();
+        new_str.push(')');
```
Comment (vil02) [inline src/backtracking/parentheses_generator.rs:42]:
```suggestion
        let new_str = current.to_string() + ")";
```

File: src/backtracking/parentheses_generator.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+/// if it's valid to do so.
+///
+/// # Arguments
+///
+/// * `current` - The current string of parentheses being built.
+/// * `open` - The count of open parentheses in the current string.
+/// * `close` - The count of close parentheses in the current string.
+/// * `n` - The total number of pairs of parentheses to be generated.
+/// * `result` - A mutable reference to the vector storing the generated combinations.
+fn generate(current: &str, open: u32, close: u32, n: u32, result: &mut Vec<String>) {
```
Comment (vil02) [inline src/backtracking/parentheses_generator.rs:28]: What do you think about renaming `open` to `open_count` etc?

File: src/backtracking/parentheses_generator.rs [hunk 4]
```diff
@@ -0,0 +1,69 @@
+/// Generates all combinations of well-formed parentheses given a positive integer `n`.
+///
+/// This function uses backtracking to generate all possible combinations of well-formed
+/// parentheses. The resulting combinations are returned as a vector of strings.
+///
+/// # Arguments
+///
+/// * `n` - A positive integer representing the number of pairs of parentheses.
+pub fn generate_parentheses(n: u32) -> Vec<String> {
```
Comment (vil02) [inline src/backtracking/parentheses_generator.rs:9]: I would use `usize` as a type of `n`, because later on, you cast it to `usize` anyway.

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/backtracking/parentheses_generator.rs (hunk 3 above)
Comment (TruongNhanNguyen) [reply src/backtracking/parentheses_generator.rs:28]: It's a good idea. I will change `open` and `close` to `open_count` and `close_count` respectively.

File: src/backtracking/parentheses_generator.rs (hunk 1 above)
Comment (TruongNhanNguyen) [reply src/backtracking/parentheses_generator.rs]: It should look like this:
Old:
```rust
let mut new_str = current.to_string();
new_str.push('(');
```
New:
```rust
let new_str = current.to_string() + "(";
```

File: src/backtracking/parentheses_generator.rs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                fn $name() {
+                    let (n, expected_result) = $test_case;
+                    assert_eq!(generate_parentheses(n), expected_result);
+                }
+            )*
+        };
+    }
+
+    generate_parentheses_tests! {
+        test_generate_parentheses_1: (1, vec!["()"]),
```
Comment (vil02) [inline src/backtracking/parentheses_generator.rs:68]: I missed that earlier: I think one should test it with `0` as an input.

I am not sure what should be the input. This has some relationship to Catalan numbers (and passes with current implementation):
```suggestion
        test_generate_parentheses_0: (0, vec![""]),
        test_generate_parentheses_1: (1, vec!["()"]),
```
This is more _intuitive_:
```suggestion
        test_generate_parentheses_0: (0, Vec::<String>::new()),
        test_generate_parentheses_1: (1, vec!["()"]),
```

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/backtracking/parentheses_generator.rs (hunk 5 above)
Comment (TruongNhanNguyen) [reply src/backtracking/parentheses_generator.rs:72]: Thanks

File: src/backtracking/parentheses_generator.rs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+///
+/// # Arguments
+///
+/// * `n` - A non-negative integer representing the number of pairs of parentheses.
+pub fn generate_parentheses(n: usize) -> Vec<String> {
+    if n == 0 {
+        return Vec::new();
+    }
+    let mut result = Vec::new();
+    generate("", 0, 0, n, &mut result);
```
Comment (vil02) [inline src/backtracking/parentheses_generator.rs:14]:
```suggestion
    let mut result = Vec::new();
    if n > 0 {
        generate("", 0, 0, n, &mut result);
    }
```

**Review by vil02 — APPROVED** [review]

**Pull request merged**